accepted/0000-install-size-info.md

```diff
@@ -0,0 +1,201 @@
+`dist.unpackedSize` and `dist.size` calculated and backfilled for all existing
+packages. This will ensure that `npm install --dry-run` can be used to do this
+kind of estimation.
+
+### npm installer changes
+
+There are two changes to be made to the installer:
+
+On the reporting side, update the three reporters (`printInstalledForHuman`,
+`printInstalledForParseable`, and `printInstalledForJSON`) in `lib/install.js`.
+You should be able to check `_size` and `_unpackedSize` in the `child.package`
+value, which holds the manifest for each installed package. Note: if **any** of
+the packages in the calculation are missing the `_unpackedSize` value and the
+current reporter is `printInstalledForHuman`, this file size summary should be
+omitted altogether. This case will happen for legacy registries when using `npm
+install --dry-run` -- it will otherwise be calculated from actual extracted
+size.
+
+On the extract side, you will need to check if the `child.package._size` and
+`child.package._unpackedSize` properties exist. If they do not, you should set
+them based on metadata returned by `pacote.extract()`.
+
+### `pacote` changes
+
+`pacote` needs to do two things to support this:
+
+* `lib/finalize-manifest.js` should be modified to get the compressed and uncompressed tarball size from package manifests, _but only if they are available_. They should be assigned to `_size` and `_unpackedSize` respectively, and are present in the `dist.size` and `dist.unpackedSize` properties of the manifest.
+
+* `extract.js` and `tarball.js` should have their functions modified such that they return or emit size and unpackedSize as part of the metadata, where appropriate, after the operation has completed. This should be calculated from the actual data getting extracted (and may require changes to `lib/extract-stream.js` as well as `lib/with-tarball-stream.js`.
+
+## Prior Art
+
+* [packagephobia.now.sh](https://packagephobia.now.sh/) -- service-based, allows pretty fast searches, but gives numbers centered more around node-side dependencies than bundle sizes -- so getting an idea about compression is impossible, and furthermore, the sizes are wildly inaccurate for a particular project because it doesn't take into account per-project deduplication. It only calculates this flattening at the level of the dependency itself. I do not believe this style of tool serves our purposes.
+* [`yarn why`](https://yarnpkg.com/en/docs/cli/why) -- this does give some level of this information, but it's not quite prepared in a way that's useful for estimating the specific impact of the package on a bundle. Furthermore, the command seems fairly broken and inaccurate as of 1.6.0 (???), so there might be something strange about how it calculates things. A nice thing about this command is that it can be used after installation to start hunting down disk-hungry packages, which is a workflow this proposal does not account for.
+
```

Comment:
It might be worth noting that [cost-of-modules](https://github.com/siddharthkp/cost-of-modules) provides a breakdown of installed package sizes. Although the focus is primarily on your already installed packages.

/cc @siddharthkp